rclcpp/src/rclcpp/parameter_client.cpp

```diff
@@ -0,0 +1,361 @@
+        callback(future_result);
+      }
+    }
+  );
+  // *INDENT-ON*
+
+  return future_result;
+}
+
+std::shared_future<rcl_interfaces::msg::SetParametersResult>
+AsyncParametersClient::set_parameters_atomically(
+  const std::vector<rclcpp::parameter::ParameterVariant> & parameters,
+  std::function<
+    void(std::shared_future<rcl_interfaces::msg::SetParametersResult>)
+  > callback)
+{
+  auto promise_result =
+    std::make_shared<std::promise<rcl_interfaces::msg::SetParametersResult>>();
+  auto future_result = promise_result->get_future().share();
+
+  auto request = std::make_shared<rcl_interfaces::srv::SetParametersAtomically::Request>();
+
+  // *INDENT-OFF* (prevent uncrustify from making unecessary indents here)
+  std::transform(parameters.begin(), parameters.end(), std::back_inserter(request->parameters),
+    [](rclcpp::parameter::ParameterVariant p) {
+      return p.to_parameter();
+    }
+  );
+
+  set_parameters_atomically_client_->async_send_request(
+    request,
+    [promise_result, future_result, &callback](
+      rclcpp::client::Client<rcl_interfaces::srv::SetParametersAtomically>::SharedFuture cb_f)
+    {
+      promise_result->set_value(cb_f.get()->result);
+      if (callback != nullptr) {
+        callback(future_result);
+      }
+    }
+  );
+  // *INDENT-ON*
+
+  return future_result;
+}
+
+std::shared_future<rcl_interfaces::msg::ListParametersResult>
+AsyncParametersClient::list_parameters(
+  const std::vector<std::string> & prefixes,
+  uint64_t depth,
+  std::function<
+    void(std::shared_future<rcl_interfaces::msg::ListParametersResult>)
+  > callback)
+{
+  auto promise_result =
+    std::make_shared<std::promise<rcl_interfaces::msg::ListParametersResult>>();
+  auto future_result = promise_result->get_future().share();
+
+  auto request = std::make_shared<rcl_interfaces::srv::ListParameters::Request>();
+  request->prefixes = prefixes;
+  request->depth = depth;
+
+  // *INDENT-OFF* (prevent uncrustify from making unecessary indents here)
+  list_parameters_client_->async_send_request(
+    request,
+    [promise_result, future_result, &callback](
+      rclcpp::client::Client<rcl_interfaces::srv::ListParameters>::SharedFuture cb_f)
+    {
+      promise_result->set_value(cb_f.get()->result);
+      if (callback != nullptr) {
+        callback(future_result);
+      }
+    }
+  );
+  // *INDENT-ON*
+
+  return future_result;
+}
+
+SyncParametersClient::SyncParametersClient(
+  rclcpp::node::Node::SharedPtr node)
+: node_(node)
+{
+  executor_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+  async_parameters_client_ = std::make_shared<AsyncParametersClient>(node);
+}
+
+SyncParametersClient::SyncParametersClient(
+  rclcpp::executor::Executor::SharedPtr executor,
+  rclcpp::node::Node::SharedPtr node)
+: executor_(executor), node_(node)
+{
+  async_parameters_client_ = std::make_shared<AsyncParametersClient>(node);
+}
+
+std::vector<rclcpp::parameter::ParameterVariant>
```

Comment:
@esteve made some changes to convert many of the vectors in these files to const references:

421f6b8b5e502fb28b7568baa26366a413db3691

it looks like they haven't been incorporated into the new cpp file, since that commit got merged after this PR was opened.